[SDESK-7285] Create Async GridFS MediaStorage client

eos87 opened the pull request:
### Purpose
Introduce the GridFSMediaStorageAsync class, which provides an asynchronous interface for storing, retrieving, and managing files in MongoDB's GridFS. The class includes methods for storing files with metadata, retrieving files by ID or filename, searching for files based on criteria, checking the existence of files, and deleting files. Documentation and usage examples have been also provided.

### What has changed
- Added GridFSMediaStorageAsync class for asynchronous file operations in GridFS.
- Implemented methods for storing, retrieving, searching, checking existence, and deleting files.
- Updated documentation with examples for each method.
- Added Sphinx documentation to include the new class and its methods.

Resolves: [SDESK-7285]

[SDESK-7285]: https://sofab.atlassian.net/browse/SDESK-7285?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

MarkLark86 commented on superdesk/storage/async_storage/media_storage.py:
```diff
@@ -0,0 +1,265 @@
+logger = logging.getLogger(__name__)
+
+
+class GridFSMediaStorageAsync(MediaStorage, MimetypeMixin):
```
We don't want to use any `eve` at all for new code. One of the goals is to completely remove Eve from all our code.

eos87 replied:
Removed inheritance from `MediaStorage`

MarkLark86 commented on superdesk/storage/async_storage/media_storage.py:
```diff
@@ -0,0 +1,265 @@
+    app: SuperdeskAsyncApp
+
+    def __init__(self, app: SuperdeskAsyncApp):
+        self.app = app
```
It would be good to not rely on any app specifics, making this storage interface generic, only relying on the PyMongo library. We can then add async GridFS functionality into the existing async mongo module (under `superdesk.core.mongo`).

eos87 replied:
done.

MarkLark86 commented on superdesk/storage/async_storage/media_storage.py:
```diff
@@ -0,0 +1,265 @@
+# -*- coding: utf-8; -*-
```
It might also be good to have this module under `superdesk.core` somewhere, to differentiate between the old and the new.

eos87 replied:
done

eos87 commented:
Hi @MarkLark86, I've implemented your suggestions and also updated [the documentation](https://superdesk--2648.org.readthedocs.build/en/2648/). Please take another look when you have some time. Thanks!

MarkLark86 commented on superdesk/core/storage.py:
```diff
@@ -0,0 +1,261 @@
+        px = mongo.get_resource_config(resource).prefix
+
+        if px not in self._fs:
+            _, db = mongo.get_client_async(resource)
```
I was thinking here, if `resource` is not registered with the new system, should we fallback to the default mongo connection config (like we're now doing with the [MongoLock](https://github.com/superdesk/superdesk-core/blob/async/superdesk/lock.py#L46))?

We could easily do this by wrapping this call into a `try...except KeyError` block, and in the exception code block get the default config using `get_mong-client_config`. This would also mean you don't need the upload resource in the tests.

Thoughts?

MarkLark86 replied:
An improvement might be to add a `get_default_client_async` function onto the [MongoResources](https://github.com/superdesk/superdesk-core/blob/async/superdesk/core/mongo.py#L172) class that will return a mongo client/db, and use that here (and in the superdesk.lock module)

petrjasek replied:
I would rather raise an error in that case, this could be error prone when you have a typo in resource you might get different connection, most of the time we get the same but in those few times when it matters it might be doing something unexpected

eos87 replied:
I think I agree with the idea of raising an error instead and let the developer handle it.

MarkLark86 reviewed (approved):
Looks good. I do have concerns around async streaming vs consuming the entire request, but that can be looked at later when we upgrade to Quart
